### ISSUE TITLE
Autogenerate alt tags for images 

### ISSUE THREAD

**Issue by KyleAMathews** [gatsbyjs/gatsby#2910]
https://twitter.com/sarah_edo/status/930083568411217920

We could add support for this to gatsby-transformer-sharp so every image includes an alt tag.

**Comment by vivek12345** [comment]
@KyleAMathews Would this be a difficult task for a newbie to take up?

**Comment by KyleAMathews** [comment]
Nope!

A PR adding this would need

* Add new field (perhaps `autoAltText`?) to https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-transformer-sharp/src/extend-node-type.js
* Add function which takes an image and first checks the cache if there's a generated alt text already and if not asks Microsoft and returns the generated text
* Add autoAltText to gatsby-image's fragments https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-transformer-sharp/src/fragments.js
* Also to https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-image/src/index.js though the code here should use a passed in alt in preference to the auto-generated one.

Do you want to take this on?

**Comment by vivek12345** [comment]
Yes I would want to take this up, but I will keep it open for other's as well if I take too much time to get this done.

**Comment by KyleAMathews** [comment]
Awesome! Let us know your progress! Also don't feel bad about throwing up a partially finished PR.

**Comment by fk** [comment]
Go @vivek12345! 👍

**Comment by vivek12345** [comment]
Thank you :)

**Comment by fk** [comment]
🙌 

Btw. [gatsby-remark-copy-linked-files](https://www.gatsbyjs.org/packages/gatsby-remark-copy-linked-files/) also generates a default alt text for HTML images in Markdown (`<img src=…>`) via the image filename here: https://github.com/gatsbyjs/gatsby/blob/d220fca29fd5d8334e5ef7eb8c48c95218dd022b/packages/gatsby-remark-copy-linked-files/src/index.js#L121-L127

AFAIU this mimics what happens to regular Markdown images that do not have an user-defined alt text (`![](../../image.jpg)`). Would be nice to somehow generalize this, too (not necessarily in one go)!

**Comment by KyleAMathews** [comment]
Yeah! Let's get this in gatsby-plugin-sharp and then add it like normal to all the usual places.

**Comment by benjaminhoffman** [comment]
When we think about the point of alt text, it's for a11y (screen readers) and SEO (tells googlebot what this is an image of).  In order of preference, it's better to have: 1) accurate, hand-typed alt text, then 2) empty string alt "", then 3) no alt text, and finally 4) incorrect alt text.

Assuming the above is true, a feature like this could cause issues b/c auto generating alt text from horribly-named images will jump to the fourth option.  If this feature gets included then maybe a warning or strong notice to the user that their images should be descriptive would be helpful.

Or instead, would it be better to check if an alt text is included and if not, add a warning informing the user that they should include an alt text in their image?

**Comment by Undistraction** [comment]
@kyleAMathews I'd definitely argue this should be built and easy to use in for the simple reason that making it easy to use will mean more people will use it, which is better for accessibility.
